README.md
# aframe-loader-3dtiles-component

[![Version](http://img.shields.io/npm/v/aframe-loader-3dtiles-component.svg?style=flat-square)](https://npmjs.org/package/aframe-loader-3dtiles-component)
[![License](http://img.shields.io/npm/l/aframe-loader-3dtiles-component.svg?style=flat-square)](https://npmjs.org/package/aframe-loader-3dtiles-component)
[![Build Status](https://cloud.drone.io/api/badges/nytimes/aframe-loader-3dtiles-component/status.svg)](https://cloud.drone.io/nytimes/aframe-loader-3dtiles-component)

An [A-Frame](https://aframe.io) component for displaying _OGC 3D Tiles_ based on [nytimes/three-loader-3dtiles](https://github.com/nytimes/three-loader-3dtiles).

[Demos](#demos) &mdash;
[API](#api) &mdash;
[Installation](#installation) &mdash;
[Usage](#usage) &mdash;
[Contributing](#contributing) &mdash;
[Roadmap](#roadmap)


<img src="https://user-images.githubusercontent.com/39962/141192128-3449b84d-8575-4e86-aab3-f615d22a16b5.jpg" width="500px">

The loader uses the [loaders.gl library](https://github.com/visgl/loaders.gl), which is part of the [vis.gl platform](https://vis.gl/), openly governed by the [Urban Computing Foundation](https://uc.foundation/). Development of the loader started at The New York Times R&D as an effort deliver massive 3D and Geographical journalism to readers. This A-Frame components makes _3D Tiles_ more accessible to creators and brings immersitve journalism into mixed realities on the web.

## Demos
### Demos are available [here](https://nytimes.github.io/aframe-loader-3dtiles-component).


### Installation

#### Browser

The following example loads the library from CDN, as well as missing required components for handling [KTX2/basis_universal](https://github.com/BinomialLLC/basis_universal) glTFs.

```html
<head>
  <title>My A-Frame Scene</title>
  <script src="https://aframe.io/releases/1.3.0/aframe.min.js"></script>
  <script src='https://cdn.jsdelivr.net/npm/three@0.137.0/examples/js/utils/WorkerPool.js'></script>
  <script src='https://cdn.jsdelivr.net/npm/three@0.137.0/examples/js/loaders/KTX2Loader.js'></script>
  <script src="https://cdn.jsdelivr.net/npm/aframe-loader-3dtiles-component/dist/aframe-loader-3dtiles-component.min.js"></script>
</head>
```
### Usage

The following example sets up a basic scene with a RealityCapture photogrammetry model.

```html
<body>
    <a-scene environment="preset: default">
      <a-camera id="camera"></a-camera>
      <a-entity
        id="freeman-tiles"
        position="0 2.6 0"
        rotation="-90 180 0"
        scale="2 2 2"
        loader-3dtiles="
         url: https://int.nyt.com/data/3dscenes/ONA360/TILESET/0731_FREEMAN_ALLEY_10M_A_36x8K__10K-PN_50P_DB/tileset_tileset.json; 
         maximumSSE: 48;
         cameraEl: #camera;
        "
      >
      </a-entity>
    </a-scene>
</body>
```

#### npm

Install via npm:

```bash
npm install aframe-loader-3dtiles-component
```

Then require and use.

```js
require('aframe');
require('aframe-loader-3dtiles-component');
```

## API

| Property | Type | Description | Default Value |
|----------|------|-------------|---------------|
| url | string | The URL of the tileset. For example if using Cesium ION, it would have the form: https://assets.cesium.com/[ASSET_ID]/tileset.json. Can be updated at runtime.| '' |
| cameraEl | selector | A selector for the camera whose position is used to determine which tiles to display at which LOD. If a `cameraEl` value is not provided then component uses `document.querySelector('a-scene').camera` to find `a-camera` primitive or entity with `camera` component. For more info see [Camera Requirement note below](#camera-requirement). The component automatically switches the camera in response to the `camera-set-active` event. | '' |
| maximumSSE | int | maximumScreenSpaceError (Optional) determines the distance from tiles in which they are refined, depending on their geometrical size. Increase the value to load lower LOD tiles from the same view distance (increases performance). **Cannot be updated at runtime.** | 16 |
| maximumMem | int | maximumMemoryUsage (Optional) determines maximum GPU memory (MB) to use for displaying tiles. May go over the limit if tiles within the camera viewport exceed that amount. **Cannot be updated at runtime.** | 32 |
| distanceScale | number | (Optional) 0-1 scale for the LOD quality. A lower value loads tiles from lower LODs (increases performance). Can be updated at runtime.| 1.0 |
| pointcloudColoring | string | (Optional) When viewing point clouds, which datum determines the color of a point. Possible values: 'white', 'intensity', 'classification', 'elevation', 'rgb'.| 'white' |
| pointcloudElevationRange | array | (Optional) When viewing point clouds colored by their elevation, what are the minimum and maximum height values.| 0, 400 |
| wireframe | boolean | (Optional) When viewing b3dm (mesh) tiles, show meshes as wireframe. Can be updated at runtime. | false | 
| showStats | boolean | (Optional) Attaches a box with tilestats to the scene, useful for debugging. | false | 
| cesiumIONToken | string | (Optional) A Cesium ION access token when loading tilesets from Cesium ION. | '' |

### Camera Requirement
To use the `loader-3dtiles` component, your A-Frame scene must have a camera defined using [`<a-camera>`](https://aframe.io/docs/1.2.0/primitives/a-camera.html) or the [`camera` component](https://aframe.io/docs/1.2.0/components/camera.html) . The component will attempt to use one of these automatically, or you can specify your own selector with the `cameraEl` property.

The camera is used to provide dynamic levels of detail (LOD) for 3D tiles. If no suitable camera is found the tiles will not display and an error is shown on the console `3D Tiles: Please add an active camera or specify the target camera via the cameraEl property`.

### Compatability with THREE.Cache
Currently, `three-loader-3dtiles` does not support the use of [THREE.Cache](https://threejs.org/docs/#api/en/loaders/Cache) which is enabled by default in A-Frame. In its default setting, the cache does not recognize repeated loads of tiles, which causes the cache to overflow. Therefore, the `loader-3dtiles` component disables `THREE.Cache` if it is enabled and displays a warning: `3D Tiles loader cannot work with THREE.Cache, disabling`.

### Contributing

Refer to [CONTRIBUTING.MD](./CONTRIBUTING.md) for general contribution instructions.


## Roadmap
Suggestions for future work include:

- More elaborate test cases.
- Better integration with the inspector.
- Test AR scenarios.
- Test point-cloud scenarios.
- Could `THREE.Cache` be used manually?
- Conversion to _Typescript_.


 ---

> This repository is maintained by the Research & Development team at The New York Times and is provided as-is for your own use. For more information about R&D at the Times visit [rd.nytimes.com](https://rd.nytimes.com)


CHANGELOG.md
## v1.0

### v1.0.6
 - Support for A-Frame 1.3/Three r137

### v1.0.4

- Bump to three-loader-3dtiles 1.1.3


tests/helpers.js
/* global suite */

/**
 * Helper method to create a scene, create an entity, add entity to scene,
 * add scene to document.
 *
 * @returns {object} An `<a-entity>` element.
 */
module.exports.entityFactory = function (opts) {
  const scene = document.createElement('a-scene');
  const assets = document.createElement('a-assets');
  const entity = document.createElement('a-entity');
  const camera = document.createElement('a-camera');

  scene.appendChild(assets);
  scene.appendChild(entity);
  scene.appendChild(camera);

  opts = opts || {};

  if (opts.assets) {
    opts.assets.forEach(function (asset) {
      assets.appendChild(asset);
    });
  }

  document.body.appendChild(scene);
  return entity;
};

/**
 * Creates and attaches a mixin element (and an `<a-assets>` element if necessary).
 *
 * @param {string} id - ID of mixin.
 * @param {object} obj - Map of component names to attribute values.
 * @param {Element} scene - Indicate which scene to apply mixin to if necessary.
 * @returns {object} An attached `<a-mixin>` element.
 */
module.exports.mixinFactory = function (id, obj, scene) {
  const mixinEl = document.createElement('a-mixin');
  mixinEl.setAttribute('id', id);
  Object.keys(obj).forEach(function (componentName) {
    mixinEl.setAttribute(componentName, obj[componentName]);
  });

  const assetsEl = scene ? scene.querySelector('a-assets') : document.querySelector('a-assets');
  assetsEl.appendChild(mixinEl);

  return mixinEl;
};

/**
 * Test that is only run locally and is skipped on CI.
 */
module.exports.getSkipCISuite = function () {
  if (window.__env__.TEST_ENV === 'ci') {
    return suite.skip;
  } else {
    return suite;
  }
};


tests/index.test.js
/* global assert, setup, suite, test */
import 'aframe';
import '../index.js';

const entityFactory = require('./helpers').entityFactory;

suite('3dtiles component', function () {
  let el;

  setup(function (done) {
    el = entityFactory();
    el.addEventListener('componentinitialized', function (evt) {
      if (evt.detail.name !== 'loader-3dtiles') { return; }
      done();
    });
    document.querySelector('a-scene').addEventListener('loaded', function () {
      el.setAttribute('loader-3dtiles', {
        url: 'https://int.nyt.com/data/3dscenes/ONA360/TILESET/0731_FREEMAN_ALLEY_10M_A_36x8K__10K-PN_50P_DB/tileset_tileset.json',
        cameraEl: '[camera]'
      });
    });
  });

  suite('model', function () {
    test('is loaded', function () {
      const tileset = document.querySelector('[loader-3dtiles]');
      tileset.addEventListener('object3dset', function (evt) {
        assert.equal(tileset.object3D.children.length, 1);
      });
    });
  });
});


tests/karma.conf.js
// Karma configuration.
process.env.CHROME_BIN = require('puppeteer').executablePath();

module.exports = function (config) {
  config.set({
    basePath: '../',
    browsers: ['ChromeHeadlessNoSandbox'],
    customLaunchers: {
      ChromeHeadlessNoSandbox: {
        base: 'ChromeHeadless',
        flags: ['--no-sandbox']
      }
    },
    client: {
      captureConsole: true,
      mocha: { ui: 'tdd' }
    },
    envPreprocessor: ['TEST_ENV'],
    files: [
      // Define test files.
      { pattern: 'tests/**/*.test.js', watched: false },
      // Serve test assets.
      { pattern: 'tests/assets/**/*', included: false, served: true }
    ],
    frameworks: ['mocha', 'sinon-chai', 'chai-shallow-deep-equal', 'webpack'],
    preprocessors: { 'tests/**/*.js': ['webpack', 'env'] },
    reporters: ['mocha'],
    webpack: {
      module: {
        rules: [
          {
            test: /\.js/,
            exclude: /(node_modules)/,
            use: ['babel-loader']
          }
        ]
      },
      externals: {
        three: 'THREE'
      }
    }
  });
};


textarea.js
if (typeof AFRAME === 'undefined') {
  throw new Error('Component attempted to register before AFRAME was available.');
}

/**
 * Textarea component for A-Frame.
 *
 * Ripped from: https://github.com/brianpeiris/aframe-textarea-component
 */

AFRAME.registerComponent('textarea', {
  schema: {
    transparentBG: { type: 'boolean', default: false },
    cols: { type: 'int', default: 40 },
    rows: { type: 'int', default: 20 },
    color: { type: 'color', default: 'black' },
    backgroundColor: { type: 'color', default: 'white' },
    disabledBackgroundColor: { type: 'color', default: 'lightgrey' },
    disabled: { type: 'boolean', default: false },
    text: { type: 'string', default: '' }
  },
  init: function () {
    this.text = null;
    this.lines = [];
    this.lastBlink = 0;
    this.blinkEnabled = !this.data.disabled;
    this.charWidth = this.charHeight = null;
    this.selectionStart = this.selectionEnd = 0;
    this.endIndexInfo = this.startIndexInfo = null;
    this.origin = { x: 0, y: 0 };

    this.background = document.createElement('a-plane');
    this.background.setAttribute('color', this.data.disabled ? this.data.disabledBackgroundColor : this.data.backgroundColor);
    this.el.appendChild(this.background);
    if (this.data.transparentBG) {
      this.background.setAttribute('material', { opacity: 0, transparent: true });
    }

    this.textAnchor = document.createElement('a-entity');
    this.el.appendChild(this.textAnchor);
    this.textAnchor.setAttribute('text', {
      mode: 'pre',
      baseline: 'top',
      anchor: 'center',
      font: 'dejavu',
      wrapCount: this.data.cols,
      height: this.data.rows,
      color: this.data.color
    });

    this._initTextarea();

    this.el.addEventListener('textfontset', this._updateCharMetrics.bind(this));
    this.el.addEventListener('char-metrics-changed', this._updateIndexInfo.bind(this));
    this.el.addEventListener('text-changed', this._updateLines.bind(this));
    this.el.addEventListener('text-changed', this._updateDisplayText.bind(this));
    this.el.addEventListener('selection-changed', this._updateIndexInfo.bind(this));
    this.el.addEventListener('selection-changed', this._updateHorizontalOrigin.bind(this));
    this.el.addEventListener('lines-changed', this._updateIndexInfo.bind(this));
    this.el.addEventListener('index-info-changed', this._updateOrigin.bind(this));
    this.el.addEventListener('index-info-changed', this._updateHorizontalOrigin.bind(this));
    this.el.addEventListener('origin-changed', this._updateDisplayText.bind(this));
    this.el.addEventListener('click', this.focus.bind(this));
  },
  update: function (oldData) {
    if (this.data.text !== oldData.text) {
      this._updateTextarea();
    }

    if ((this.data.backgroundColor !== oldData.backgroundColor || this.data.disabledBackgroundColor !== oldData.disabledBackgroundColor)) {
      this.background.setAttribute('color', this.data.disabled ? this.data.disabledBackgroundColor : this.data.backgroundColor);
    }

    if (this.data.disabled !== oldData.disabled) {
      this.blinkEnabled = !this.data.disabled;
      this.textarea.disabled = this.data.disabled;
      this.background.setAttribute('color', this.data.disabled ? this.data.disabledBackgroundColor : this.data.backgroundColor);
    }
  },
  focus: function () {
    this.textarea.focus();
  },
  _initTextarea: function () {
    this.textarea = document.createElement('textarea');
    document.body.appendChild(this.textarea);
    this._updateTextarea();
  },
  _updateTextarea: function () {
    this.textarea.style.whiteSpace = 'pre';
    this.textarea.style.overflow = 'hidden';
    this.textarea.style.opacity = '0';

    this.textarea.cols = this.data.cols;
    this.textarea.rows = this.data.rows;
    this.textarea.value = this.data.text;
    this.textarea.selectionStart = 0;
    this.textarea.selectionEnd = 0;

    this._updateIndexInfo();
  },
  _emit: function (eventName, detail) {
    this.el.emit(eventName, detail);
  },
  _updateCharMetrics: function (event) {
    const layout = this.textAnchor.components.text.geometry.layout;
    const fontWidthFactor = event.detail.fontObj.widthFactor;
    this.charWidth = fontWidthFactor * this.textAnchor.object3DMap.text.scale.x;
    this.charHeight = this.charWidth * layout.lineHeight / fontWidthFactor;
    this.textAnchor.setAttribute('position', { x: 0, y: this.charHeight * this.data.rows / 2, z: 0 });
    if (!this.data.transparentBG) {
      this.background.setAttribute('scale', { x: 1.05, y: this.charHeight * this.data.rows * 1.05, z: 1 });
      this.background.setAttribute('position', { x: 0, y: 0, z: 0 });
    }

    this._emit('char-metrics-changed');
  },
  _checkAndUpdateSelection: function () {
    if (
      this.selectionStart === this.textarea.selectionStart &&
            this.selectionEnd === this.textarea.selectionEnd
    ) {
      return;
    }

    const lastStart = this.selectionStart;
    const lastEnd = this.selectionEnd;

    this.selectionStart = this.textarea.selectionStart;
    this.selectionEnd = this.textarea.selectionEnd;

    this._emit('selection-changed', {
      start: { old: lastStart, new: this.selectionStart, changed: this.selectionStart !== lastStart },
      end: { old: lastEnd, new: this.selectionEnd, changed: this.selectionEnd !== lastEnd }
    });
  },
  tick: function (time) {
    if (time - this.lastBlink > 500 && this.blinkEnabled) {
      this.lastBlink = time;
    }
    this._checkAndUpdateSelection();
    this._checkAndUpdateText();
  },
  _getIndexInfo: function (lineIndex, textIndex) {
    const y = Math.max(0, lineIndex);
    const line = this.lines[y];
    const x = textIndex - line.start;
    return {
      line: line,
      x: x * this.charWidth,
      y: -this.charHeight * y + -this.charHeight / 2
    };
  },
  _updateIndexInfo: function () {
    if (!this.lines.length) {
      return;
    }
    const lastStart = this.startIndexInfo && this.startIndexInfo.line.index;
    const lastEnd = this.endIndexInfo && this.endIndexInfo.line.index;
    this.startIndexInfo = null;
    this.endIndexInfo = null;
    let i;
    let startChanged = false;
    let endChanged = false;
    for (i = 0; i <= this.lines.length; i++) {
      const prevLine = this.lines[i - 1];
      const lineStart = i === this.lines.length ? (prevLine.start + prevLine.length + 1) : this.lines[i].start;
      if (lineStart > this.selectionStart && !this.startIndexInfo) {
        this.startIndexInfo = this._getIndexInfo(i - 1, this.selectionStart);
        if (this.startIndexInfo.line.index !== lastStart) {
          startChanged = true;
        }
      }
      if (lineStart > this.selectionEnd) {
        this.endIndexInfo = this._getIndexInfo(i - 1, this.selectionEnd);
        if (this.endIndexInfo.line.index !== lastEnd) {
          endChanged = true;
        }
        break;
      }
    }
    if (startChanged || endChanged) {
      this._emit('index-info-changed', {
        start: { changed: startChanged },
        end: { changed: endChanged }
      });
    }
  },
  _updateOrigin: function (event) {
    let changed = false;
    if (event.detail.end.changed) {
      const end = this.origin.y + this.data.rows - 1;
      if (this.endIndexInfo.line.index > end) {
        this.origin.y = this.endIndexInfo.line.index + 1 - this.data.rows;
        changed = true;
      } else if (this.endIndexInfo.line.index < this.origin.y) {
        this.origin.y = this.endIndexInfo.line.index;
        changed = true;
      }
    }
    if (event.detail.start.changed) {
      if (this.startIndexInfo.line.index < this.origin.y) {
        this.origin.y = this.startIndexInfo.line.index;
        changed = true;
      }
    }
    if (changed) {
      this._emit('origin-changed');
    }
  },
  _updateHorizontalOrigin: function (event) {
    if (!this.endIndexInfo) {
      return;
    }
    let changed = true;
    if (event.detail.end.changed) {
      const endIndex = this.selectionEnd - this.endIndexInfo.line.start;
      if (endIndex > this.origin.x + this.data.cols) {
        this.origin.x = endIndex - this.data.cols;
        changed = true;
      } else if (endIndex < this.origin.x) {
        this.origin.x = endIndex;
        changed = true;
      }
    }
    const startIndex = this.selectionStart - this.startIndexInfo.line.start;
    if (event.detail.start.changed) {
      if (startIndex > this.origin.x + this.data.cols) {
        this.origin.x = startIndex - this.data.cols;
        changed = true;
      } else if (startIndex < this.origin.x) {
        this.origin.x = startIndex;
        changed = true;
      }
    }
    if (changed) {
      this._emit('origin-changed');
    }
  },
  _updateLines: function () {
    this.lines = [];
    const lines = this.text.split('\n');
    let counter = 0;
    for (let i = 0; i < lines.length; i++) {
      this.lines[i] = {
        index: i,
        length: lines[i].length,
        start: counter
      };
      counter += lines[i].length + 1;
    }
    this._emit('lines-changed');
  },
  _getViewportText: function () {
    return this.text.split('\n').slice(this.origin.y, this.origin.y + this.data.rows)
      .map(function (line) {
        return line.substr(this.origin.x, this.data.cols) || ' ';
      }.bind(this)).join('\n');
  },
  _updateDisplayText: function () {
    this.textAnchor.setAttribute('text', {
      value: this._getViewportText()
    });
  },
  _checkAndUpdateText: function () {
    const text = this.textarea.value;
    if (text === this.text) {
      return;
    }
    this.text = text;
    this._emit('text-changed');
  }
});


webpack.config.js
const path = require('path');

const PLUGINS = [];

module.exports = {
  devServer: {
    static: __dirname,
    host: '0.0.0.0'
  },
  entry: './index.js',
  mode: process.env.NODE_ENV === 'production' ? 'production' : 'development',
  output: {
    globalObject: 'this',
    path: path.join(__dirname, 'dist'),
    filename: process.env.NODE_ENV === 'production' ? 'aframe-loader-3dtiles-component.min.js' : 'aframe-loader-3dtiles-component.js',
    publicPath: '/dist',
    libraryTarget: 'umd'
  },
  plugins: PLUGINS,
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /(node_modules)/,
        use: ['babel-loader']
      }
    ]
  },
  resolve: {
    extensions: ['.js'],
    modules: [path.join(__dirname, 'node_modules')]
  },
  externals: {
    three: 'THREE',
    'three/examples/jsm/loaders/GLTFLoader.js': 'THREE',
    'three/examples/jsm/loaders/DRACOLoader.js': 'THREE',
    'three/examples/jsm/loaders/KTX2Loader.js': 'THREE'
  }
};


index.js
import { Loader3DTiles, PointCloudColoring } from 'three-loader-3dtiles';
import { Vector3 } from 'three';
import './textarea';

if (typeof AFRAME === 'undefined') {
  throw new Error('Component attempted to register before AFRAME was available.');
}

const POINT_CLOUD_COLORING = {
  white: PointCloudColoring.White,
  intensity: PointCloudColoring.Intensity,
  classification: PointCloudColoring.Classification,
  elevation: PointCloudColoring.Elevation,
  rgb: PointCloudColoring.RGB
};
/**
 * 3D Tiles component for A-Frame.
 */

AFRAME.registerComponent('loader-3dtiles', {
  schema: {
    url: { type: 'string' },
    cameraEl: { type: 'selector' },
    maximumSSE: { type: 'int', default: 16 },
    maximumMem: { type: 'int', default: 32 },
    distanceScale: { type: 'number', default: 1.0 },
    pointcloudColoring: { type: 'string', default: 'white' },
    pointcloudElevationRange: { type: 'array', default: ['0', '400'] },
    wireframe: { type: 'boolean', default: false },
    showStats: { type: 'boolean', default: false },
    cesiumIONToken: { type: 'string' }
  },
  init: async function () {
    this.camera = this.data.cameraEl?.object3D.children[0] ?? document.querySelector('a-scene').camera;
    if (!this.camera) {
      throw new Error('3D Tiles: Please add an active camera or specify the target camera via the cameraEl property');
    }
    const { model, runtime } = await this._initTileset();
    this.el.setObject3D('tileset', model);

    this.originalCamera = this.camera;
    this.el.sceneEl.addEventListener('camera-set-active', (e) => {
      // TODO: For some reason after closing the inspector this event is fired with an empty camera,
      // so revert to the original camera used.
      //
      // TODO: Does not provide the right Inspector perspective camera
      this.camera = e.detail.cameraEl.object3D.children[0] ?? this.originalCamera;
    });
    this.el.sceneEl.addEventListener('enter-vr', (e) => {
      this.originalCamera = this.camera;
      try {
        this.camera = this.el.sceneEl.renderer.xr.getCamera(this.camera);

        // FOV Code from https://github.com/mrdoob/three.js/issues/21869
        this.el.sceneEl.renderer.xr.getSession().requestAnimationFrame((time, frame) => {
          const ref = this.el.sceneEl.renderer.xr.getReferenceSpace();
          const pose = frame.getViewerPose(ref);
          if (pose) {
            const fovi = pose.views[0].projectionMatrix[5];
            this.camera.fov = Math.atan2(1, fovi) * 2 * 180 / Math.PI;
          }
        });
      } catch (e) {
        console.warn('Could not get VR camera');
      }
    });
    this.el.sceneEl.addEventListener('exit-vr', (e) => {
      this.camera = this.originalCamera;
    });

    if (this.data.showStats) {
      this.stats = this._initStats();
    }
    if (THREE.Cache.enabled) {
      console.warn('3D Tiles loader cannot work with THREE.Cache, disabling.');
      THREE.Cache.enabled = false;
    }
    await this._nextFrame();
    this.runtime = runtime;
    this.runtime.setElevationRange(this.data.pointcloudElevationRange.map(n => Number(n)));
  },
  update: async function (oldData) {
    if (oldData.url !== this.data.url) {
      if (this.runtime) {
        this.runtime.dispose();
        this.runtime = null;
      }
      const { model, runtime } = await this._initTileset();
      this.el.setObject3D('tileset', model);
      await this._nextFrame();
      this.runtime = runtime;
    } else if (this.runtime) {
      this.runtime.setPointCloudColoring(this._resolvePointcloudColoring(this.data.pointCloudColoring));
      this.runtime.setWireframe(this.data.wireframe);
      this.runtime.setViewDistanceScale(this.data.distanceScale);
      this.runtime.setElevationRange(this.data.pointcloudElevationRange.map(n => Number(n)));
    }

    if (this.data.showStats && !this.stats) {
      this.stats = this._initStats();
    }
    if (!this.data.showStats && this.stats) {
      this.el.sceneEl.removeChild(this.stats);
      this.stats = null;
    }
  },
  tick: function (t, dt) {
    if (this.runtime) {
      this.runtime.update(dt, this.el.sceneEl.renderer, this.camera);
      if (this.stats) {
        const worldPos = new Vector3();
        this.camera.getWorldPosition(worldPos);
        const stats = this.runtime.getStats();
        this.stats.setAttribute(
          'textarea',
          'text',
          Object.values(stats.stats).map(s => `${s.name}: ${s.count}`).join('\n')
        );
        const newPos = new Vector3();
        newPos.copy(worldPos);
        newPos.z -= 2;
        this.stats.setAttribute('position', newPos);
      }
    }
  },
  remove: function () {
    if (this.runtime) {
      this.runtime.dispose();
    }
  },
  _resolvePointcloudColoring () {
    const pointCloudColoring = POINT_CLOUD_COLORING[this.data.pointcloudColoring];
    if (!pointCloudColoring) {
      console.warn('Invalid value for point cloud coloring');
      return PointCloudColoring.White;
    } else {
      return pointCloudColoring;
    }
  },
  _initTileset: async function () {
    const pointCloudColoring = this._resolvePointcloudColoring(this.data.pointcloudColoring);

    return Loader3DTiles.load({
      url: this.data.url,
      renderer: this.el.sceneEl.renderer,
      options: {
        dracoDecoderPath: 'https://unpkg.com/three@0.137.0/examples/js/libs/draco',
        basisTranscoderPath: 'https://unpkg.com/three@0.137.0/examples/js/libs/basis',
        cesiumIONToken: this.data.cesiumIONToken,
        maximumScreenSpaceError: this.data.maximumSSE,
        maximumMemoryUsage: this.data.maximumMem,
        viewDistanceScale: this.data.distanceScale,
        wireframe: this.data.wireframe,
        pointCloudColoring: pointCloudColoring,
        updateTransforms: true
      }
    });
  },
  _initStats: function () {
    const stats = document.createElement('a-entity');
    this.el.sceneEl.appendChild(stats);
    stats.setAttribute('position', '-0.5 0 -1');
    stats.setAttribute('textarea', {
      cols: 30,
      rows: 15,
      text: '',
      color: 'white',
      disabledBackgroundColor: '#0c1e2c',
      disabled: true
    });
    return stats;
  },
  _nextFrame: async function () {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve();
      }, 0);
    });
  }
});
